### PR TITLE
fix #174 Drop mongodb roles index "name_1" in _forwardMigrate

### LINKED ISSUE #174: [v2.0] Index error when migrating
In my mongodb console : `db.roles.getIndexes()` prints

``` json
[
	{
		"v" : 1,
		"key" : {
			"_id" : 1
		},
		"name" : "_id_",
		"ns" : "admin.roles"
	},
	{
		"v" : 1,
		"unique" : true,
		"key" : {
			"name" : 1
		},
		"name" : "name_1",
		"ns" : "admin.roles"
	}
]
```
This prompt an error at meteor startup : `MongoError: E11000 duplicate key error index: admin.roles.$name_1 dup key: { : null }` because field `name` isn't used anymore in v2.

This error is triggered despite the running of `Package['alanning:roles'].Roles._forwardMigrate()` from the meteor shell. 

Had to run `db.roles.dropIndex("name_1")` from the mongo shell!
Hope this helps, cheers :)

### PR DESCRIPTION
**Pull request by sveinburne**
See #174

**WARNING** : throws an exception if index "name_1" does not exists. Any reason to think this might happen? 
In that case, might want to surround it with a try catch block and print a warning.

### CODE REVIEW

File: roles/roles_server.js [hunk 1]
```diff
@@ -248,6 +248,9 @@ _.extend(Roles, {
     updateUser = updateUser || Roles._defaultUpdateUser;
     updateRole = updateRole || Roles._defaultUpdateRole;
 
+    //drop old index, preventing mongo duplicate key 'null' errors
```
Comment (mitar) [inline roles/roles_server.js]: Please keep comments in style with other comments.

**Comment by mitar** [comment]
Thanks for the pull request.

I would wrap it into the try/catch, but also verify that the error thrown is really that the index does not exist. Otherwise rethrow the exception.

Also, do the inverse for backwards migration.

**Comment by sveinburne** [comment]
There is unfortunately no neat way to check against this peculiar exception. I have made up this workaround though : 

```javascript
    try {
      Meteor.roles._dropIndex("name_1");
    } catch (error) {
      if(error.errmsg.match(/index not found/)) {
        console.warn("An index 'name_1' from v1 release in roles collection was expected to be drop");
      } else {
        throw error;
      }
    }
```
But I'm not certain this message doesn't depend on mongodb driver or mongodb release. 
**PS** : 
* full message is `index not found with name [name_1]`
* Meteor 1.2.1
* mongodb v3.0.7

**Comment by mitar** [comment]
See [code example here](https://github.com/alanning/meteor-roles/blob/v2.0/roles/roles_common.js#L79). So maybe do a similar check for `MongoError`.

And do not worry about warning. Just ignore in that case. We just want to prevent errors later on. We do not really care if index is there. We just do not want to interfere with it later.

Also respect code style.

**Comment by sveinburne** [comment]
Ok I updated the above snippet, hope code style is fitting now. But something is intriguing me : in the portion of code you referred, you used `e.err` while I had to use `e.errmsg` (`e.err` was `undefined`), any thoughts ?

**Comment by mitar** [comment]
Interesting. Even in one [other code](https://github.com/mitar/council-app/blob/96901bb40dc6c81e04a7d285ef23a443ef250547/packages/api/motion/methods.coffee#L285) I use `e.err`. Not sure where `errmsg` is coming from. You could turn around your regex as I have in this code and then do `(e.err || e.errmsg)` just to be sure.

You can fix the same also in the other part of the code.

And also make removing/adding indexes for forward migration.

And just work in the pull request, not by editing the code in the comment.

**Comment by sveinburne** [comment]
Yeah well that was to prevent from having 10 commits instead of 2 sensible ones!
I'll do the forward thing during next week, and commit  this snippet + err | errmsg changes tonight.

**Comment by mitar** [comment]
> Yeah well that was to prevent from having 10 commits instead of 2 sensible ones!

You can always squash them later.

**Comment by sveinburne** [comment]
Done !

**Comment by mitar** [comment]
Style is a bit off. When you are contributing to an existing project, always try to match code style of the rest of the code. And in general be mindful of your code style. Even your own code style of code you are contributing is not consistent.

File: roles/roles_server.js [hunk 2]
```diff
@@ -5,9 +5,25 @@
...
+ *
+ * @param {Object} collection to be proceeded
+ * @param {String} indexName to drop
+ */
+dropCollectionIndex = function(collection,indexName){
+  try {
+    collection._dropIndex(indexName);
+  } catch (e) {
+    if (e.name !== 'MongoError') throw e;
+    match = (e.err || e.errmsg).match(/index not found/);
```
Comment (mitar) [inline roles/roles_server.js]: This variable is missing `var` to make it local.

**Comment by mitar** [comment]
Fixed style in 76fe6541573e77d1f6cb2ab5875c38894a4a40dc.

Thanks for reporting this and fixing it.

File: roles/roles_server.js (hunk 2 above)
Comment (sveinburne) [inline roles/roles_server.js]: Oops thanks for reporting, too much coffeescript in the last months I guess...

**Comment by sveinburne** [comment]
Thanks mitar for your patience and advices. I will pay more attention to code style next time I contribute.

**Pull request merged**